core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```
@@ -0,0 +1,218 @@
+          .order(ByteOrder.LITTLE_ENDIAN).getInt();
+      Preconditions.checkState(plainBlockSize > 0, "Wrong plainBlockSize " + plainBlockSize);
+
+      cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
+      this.ciphertextBlockBuffer = new byte[cipherBlockSize];
+      this.currentBlockIndex = 0;
+      this.currentOffsetInPlainBlock = 0;
+
+      int numberOfFullBlocks = Math.toIntExact(netSourceFileSize / cipherBlockSize);
+      int cipherBytesInLastBlock = Math.toIntExact(netSourceFileSize - numberOfFullBlocks * cipherBlockSize);
+      boolean fullBlocksOnly = (0 == cipherBytesInLastBlock);
+      numberOfBlocks = fullBlocksOnly ? numberOfFullBlocks : numberOfFullBlocks + 1;
+      lastCipherBlockSize = fullBlocksOnly ? cipherBlockSize : cipherBytesInLastBlock; // never 0
+      int plainBytesInLastBlock = fullBlocksOnly ? 0 :
+          (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);
+      plainStreamSize = numberOfFullBlocks * plainBlockSize + plainBytesInLastBlock;
+    } else {
+      plainStreamSize = 0;
+
+      gcmDecryptor = null;
+      ciphertextBlockBuffer = null;
+      cipherBlockSize = -1;
+      plainBlockSize = -1;
+      numberOfBlocks = -1;
+      lastCipherBlockSize = -1;
+      this.fileAadPrefix = null;
+    }
+  }
+
+  public long plaintextStreamSize() {
+    return plainStreamSize;
+  }
+
+  @Override
+  public int available() throws IOException {
+    long maxAvailable = plainStreamSize - plainStreamPosition;
+    // See InputStream.available contract
+    if (maxAvailable >= Integer.MAX_VALUE) {
+      return Integer.MAX_VALUE;
+    } else {
+      return (int) maxAvailable;
+    }
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (len < 0) {
+      throw new IOException("Negative read length " + len);
+    }
+
+    if (available() <= 0) {
+      return -1;
```

I think most Iceberg streams throw `EOFException` instead of returning -1.

---

I see, lets do the same here.